snap/snapshots.go

```diff
@@ -27,27 +27,104 @@ import (
 	"path/filepath"
 	"strings"
 
-	"gopkg.in/yaml.v2"
-
 	"github.com/snapcore/snapd/strutil"
+	"gopkg.in/yaml.v2"
 )
 
 var osOpen = os.Open
 
-// SnapshotOptions contains the information provided by a snap package about
-// what files should be included in the snapshots.
+// SnapshotOptions describes the options available for snapshots.
+// The initial source of these options is a file in the snap package.
+// In addition, options can be modified with dynamic requests via REST API.
 type SnapshotOptions struct {
 	// ExcludePaths is the list of file and directory patterns that need to be
 	// excluded from a snapshot. At the moment the only supported globbing
 	// character is "*", which stands for any sequence of characters other than
 	// "/".
-	ExcludePaths []string `yaml:"exclude"`
+	ExcludePaths []string `yaml:"exclude" json:"exclude,omitempty"`
 }
 
 const (
 	snapshotManifestPath = "meta/snapshots.yaml"
 )
 
+func (opts *SnapshotOptions) validateExcludePaths() error {
+	if err := validateExcludePaths(opts.ExcludePaths); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (opts *SnapshotOptions) mergeExcludePaths(paths []string) error {
+	if err := validateExcludePaths(paths); err != nil {
+		return err
+	}
+	opts.ExcludePaths = append(opts.ExcludePaths, paths...)
+
+	return nil
+}
+
+// IsEmpty determines if SnapshotOptions structure is empty.
+func (opts *SnapshotOptions) IsEmpty() bool {
+	if len(opts.ExcludePaths) == 0 {
+		return true
+	}
+
+	return false
+}
+
+// Validate validates all options.
+func (opts *SnapshotOptions) Validate() error {
+	if err := opts.validateExcludePaths(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Merge combines existing with additional options.
+func (opts *SnapshotOptions) Merge(moreOptions *SnapshotOptions) error {
+	if moreOptions != nil {
```

Comment:
Maybe: `if moreOptions == nil { return nil }` here? And then there is no need to nest deeper.

Reply:
My engraved C world pattern to avoid multiple returns wants to pop out :)
Improved.